// Data.Tuple.Nested/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Tuple = require("Data.Tuple");
var $div$bslash = function (a) {
    return function (b) {
        return new Data_Tuple.Tuple(a, b);
    };
};
var con9 = function (f) {
    return function (_199) {
        return f(_199.value0)(_199.value1.value0)(_199.value1.value1.value0)(_199.value1.value1.value1.value0)(_199.value1.value1.value1.value1.value0)(_199.value1.value1.value1.value1.value1.value0)(_199.value1.value1.value1.value1.value1.value1.value0)(_199.value1.value1.value1.value1.value1.value1.value1.value0)(_199.value1.value1.value1.value1.value1.value1.value1.value1);
    };
};
var con8 = function (f) {
    return function (_198) {
        return f(_198.value0)(_198.value1.value0)(_198.value1.value1.value0)(_198.value1.value1.value1.value0)(_198.value1.value1.value1.value1.value0)(_198.value1.value1.value1.value1.value1.value0)(_198.value1.value1.value1.value1.value1.value1.value0)(_198.value1.value1.value1.value1.value1.value1.value1);
    };
};
var con7 = function (f) {
    return function (_197) {
        return f(_197.value0)(_197.value1.value0)(_197.value1.value1.value0)(_197.value1.value1.value1.value0)(_197.value1.value1.value1.value1.value0)(_197.value1.value1.value1.value1.value1.value0)(_197.value1.value1.value1.value1.value1.value1);
    };
};
var con6 = function (f) {
    return function (_196) {
        return f(_196.value0)(_196.value1.value0)(_196.value1.value1.value0)(_196.value1.value1.value1.value0)(_196.value1.value1.value1.value1.value0)(_196.value1.value1.value1.value1.value1);
    };
};
var con5 = function (f) {
    return function (_195) {
        return f(_195.value0)(_195.value1.value0)(_195.value1.value1.value0)(_195.value1.value1.value1.value0)(_195.value1.value1.value1.value1);
    };
};
var con4 = function (f) {
    return function (_194) {
        return f(_194.value0)(_194.value1.value0)(_194.value1.value1.value0)(_194.value1.value1.value1);
    };
};
var con3 = function (f) {
    return function (_193) {
        return f(_193.value0)(_193.value1.value0)(_193.value1.value1);
    };
};
var con2 = function (f) {
    return function (_192) {
        return f(_192.value0)(_192.value1);
    };
};
var con10 = function (f) {
    return function (_200) {
        return f(_200.value0)(_200.value1.value0)(_200.value1.value1.value0)(_200.value1.value1.value1.value0)(_200.value1.value1.value1.value1.value0)(_200.value1.value1.value1.value1.value1.value0)(_200.value1.value1.value1.value1.value1.value1.value0)(_200.value1.value1.value1.value1.value1.value1.value1.value0)(_200.value1.value1.value1.value1.value1.value1.value1.value1.value0)(_200.value1.value1.value1.value1.value1.value1.value1.value1.value1);
    };
};
module.exports = {
    "/\\": $div$bslash, 
    con10: con10, 
    con9: con9, 
    con8: con8, 
    con7: con7, 
    con6: con6, 
    con5: con5, 
    con4: con4, 
    con3: con3, 
    con2: con2
};
