// Data.Foreign/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Function = require("Data.Function");
var Data_Array = require("Data.Array");
var Data_Either = require("Data.Either");

  function parseJSONImpl(left, right, str) {
    try {
      return right(JSON.parse(str));
    } catch (e) {
      return left(e.toString());
    }
  }
  ;

  function toForeign(value) {
    return value;
  }
  ;

  function unsafeFromForeign(value) {
    return value;
  }
  ;

  function typeOf(value) {
    return typeof value;
  }
  ;

  function tagOf(value) {
    return Object.prototype.toString.call(value).slice(8, -1);
  }
  ;

  function isNull(value) {
    return value === null;
  }
  ;

  function isUndefined(value) {
    return value === undefined;
  }
  ;

  var isArray = Array.isArray || function(value) {
    return Object.prototype.toString.call(value) === '[object Array]';
  };
  ;
var TypeMismatch = (function () {
    function TypeMismatch(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    TypeMismatch.create = function (value0) {
        return function (value1) {
            return new TypeMismatch(value0, value1);
        };
    };
    return TypeMismatch;
})();
var ErrorAtIndex = (function () {
    function ErrorAtIndex(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    ErrorAtIndex.create = function (value0) {
        return function (value1) {
            return new ErrorAtIndex(value0, value1);
        };
    };
    return ErrorAtIndex;
})();
var ErrorAtProperty = (function () {
    function ErrorAtProperty(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    ErrorAtProperty.create = function (value0) {
        return function (value1) {
            return new ErrorAtProperty(value0, value1);
        };
    };
    return ErrorAtProperty;
})();
var JSONError = (function () {
    function JSONError(value0) {
        this.value0 = value0;
    };
    JSONError.create = function (value0) {
        return new JSONError(value0);
    };
    return JSONError;
})();
var unsafeReadPrim = function (_97) {
    return function (_98) {
        if (tagOf(_98) === _97) {
            return Prelude.pure(Data_Either.applicativeEither)(unsafeFromForeign(_98));
        };
        return new Data_Either.Left(new TypeMismatch(_97, tagOf(_98)));
    };
};
var showForeignError = new Prelude.Show(function (_100) {
    if (_100 instanceof TypeMismatch) {
        return "Type mismatch: expected " + (_100.value0 + (", found " + _100.value1));
    };
    if (_100 instanceof ErrorAtIndex) {
        return "Error at array index " + (Prelude.show(Prelude.showNumber)(_100.value0) + (": " + Prelude.show(showForeignError)(_100.value1)));
    };
    if (_100 instanceof ErrorAtProperty) {
        return "Error at property " + (Prelude.show(Prelude.showString)(_100.value0) + (": " + Prelude.show(showForeignError)(_100.value1)));
    };
    if (_100 instanceof JSONError) {
        return "JSON error: " + _100.value0;
    };
    throw new Error("Failed pattern match");
});
var readString = unsafeReadPrim("String");
var readNumber = unsafeReadPrim("Number");
var readBoolean = unsafeReadPrim("Boolean");
var readArray = function (_99) {
    if (isArray(_99)) {
        return Prelude.pure(Data_Either.applicativeEither)(unsafeFromForeign(_99));
    };
    return new Data_Either.Left(new TypeMismatch("array", tagOf(_99)));
};
var parseJSON = function (json) {
    return parseJSONImpl(Prelude["<<<"](Prelude.semigroupoidArr)(Data_Either.Left.create)(JSONError.create), Data_Either.Right.create, json);
};
var eqForeignError = new Prelude.Eq(function (a) {
    return function (b) {
        return !Prelude["=="](eqForeignError)(a)(b);
    };
}, function (_101) {
    return function (_102) {
        if (_101 instanceof TypeMismatch && _102 instanceof TypeMismatch) {
            return _101.value0 === _102.value0 && _101.value1 === _102.value1;
        };
        if (_101 instanceof ErrorAtIndex && _102 instanceof ErrorAtIndex) {
            return _101.value0 === _102.value0 && Prelude["=="](eqForeignError)(_101.value1)(_102.value1);
        };
        if (_101 instanceof ErrorAtProperty && _102 instanceof ErrorAtProperty) {
            return _101.value0 === _102.value0 && Prelude["=="](eqForeignError)(_101.value1)(_102.value1);
        };
        if (_101 instanceof JSONError && _102 instanceof JSONError) {
            return _101.value0 === _102.value0;
        };
        return false;
    };
});
module.exports = {
    TypeMismatch: TypeMismatch, 
    ErrorAtIndex: ErrorAtIndex, 
    ErrorAtProperty: ErrorAtProperty, 
    JSONError: JSONError, 
    readArray: readArray, 
    readNumber: readNumber, 
    readBoolean: readBoolean, 
    readString: readString, 
    isArray: isArray, 
    isUndefined: isUndefined, 
    isNull: isNull, 
    tagOf: tagOf, 
    typeOf: typeOf, 
    unsafeFromForeign: unsafeFromForeign, 
    toForeign: toForeign, 
    parseJSON: parseJSON, 
    showForeignError: showForeignError, 
    eqForeignError: eqForeignError
};
