// Data.DOM.Simple.Ajax/index.js
"use strict";
var Data_Function = require("Data.Function");
var Prelude = require("Prelude");
var Data_DOM_Simple_Unsafe_Ajax = require("Data.DOM.Simple.Unsafe.Ajax");
var Control_Monad_Eff = require("Control.Monad.Eff");
var Data_Maybe = require("Data.Maybe");
var Data_DOM_Simple_Types = require("Data.DOM.Simple.Types");

  function maybeFn(nothing, just, a) {
    return a == null ? nothing : just(a);
  };

  function makeXMLHttpRequest() {
    return new XMLHttpRequest();
  };

  function responseText(obj) {
      return function () {
        return obj.responseText;
    };
  };

  function status(obj) {
    return function () {
      return obj.status;
    };
  };

  function statusText(obj) {
      return function () {
        return obj.statusText;
    };
  };

  function setRequestHeader(key) {
     return function (value) {
       return function (obj) {
         return function () {
           obj.setRequestHeader(key, value);
           return {};
         };
       };
     };
  };

  function getAllResponseHeaders(obj) {
    return function () {
      return obj.getAllResponseHeaders();
    };
  };

  function overrideMimeType(mime) {
    return function (obj) {
      return function () {
        obj.overrideMimeType(mine);
        return {};
      };
    };
  };
var Default = (function () {
    function Default() {

    };
    Default.value = new Default();
    return Default;
})();
var ArrayBuffer = (function () {
    function ArrayBuffer() {

    };
    ArrayBuffer.value = new ArrayBuffer();
    return ArrayBuffer;
})();
var Blob = (function () {
    function Blob() {

    };
    Blob.value = new Blob();
    return Blob;
})();
var Document = (function () {
    function Document() {

    };
    Document.value = new Document();
    return Document;
})();
var Json = (function () {
    function Json() {

    };
    Json.value = new Json();
    return Json;
})();
var Text = (function () {
    function Text() {

    };
    Text.value = new Text();
    return Text;
})();
var MozBlob = (function () {
    function MozBlob() {

    };
    MozBlob.value = new MozBlob();
    return MozBlob;
})();
var MozChunkedText = (function () {
    function MozChunkedText() {

    };
    MozChunkedText.value = new MozChunkedText();
    return MozChunkedText;
})();
var MozChunkedArrayBuffer = (function () {
    function MozChunkedArrayBuffer() {

    };
    MozChunkedArrayBuffer.value = new MozChunkedArrayBuffer();
    return MozChunkedArrayBuffer;
})();
var Unsent = (function () {
    function Unsent() {

    };
    Unsent.value = new Unsent();
    return Unsent;
})();
var Opened = (function () {
    function Opened() {

    };
    Opened.value = new Opened();
    return Opened;
})();
var HeadersReceived = (function () {
    function HeadersReceived() {

    };
    HeadersReceived.value = new HeadersReceived();
    return HeadersReceived;
})();
var Loading = (function () {
    function Loading() {

    };
    Loading.value = new Loading();
    return Loading;
})();
var Done = (function () {
    function Done() {

    };
    Done.value = new Done();
    return Done;
})();
var GET = (function () {
    function GET() {

    };
    GET.value = new GET();
    return GET;
})();
var POST = (function () {
    function POST() {

    };
    POST.value = new POST();
    return POST;
})();
var PUT = (function () {
    function PUT() {

    };
    PUT.value = new PUT();
    return PUT;
})();
var DELETE = (function () {
    function DELETE() {

    };
    DELETE.value = new DELETE();
    return DELETE;
})();
var PATCH = (function () {
    function PATCH() {

    };
    PATCH.value = new PATCH();
    return PATCH;
})();
var HEAD = (function () {
    function HEAD() {

    };
    HEAD.value = new HEAD();
    return HEAD;
})();
var OPTIONS = (function () {
    function OPTIONS() {

    };
    OPTIONS.value = new OPTIONS();
    return OPTIONS;
})();
var JSONP = (function () {
    function JSONP() {

    };
    JSONP.value = new JSONP();
    return JSONP;
})();
var HttpMethod = (function () {
    function HttpMethod(value0) {
        this.value0 = value0;
    };
    HttpMethod.create = function (value0) {
        return new HttpMethod(value0);
    };
    return HttpMethod;
})();
var NoData = (function () {
    function NoData() {

    };
    NoData.value = new NoData();
    return NoData;
})();
var TextData = (function () {
    function TextData(value0) {
        this.value0 = value0;
    };
    TextData.create = function (value0) {
        return new TextData(value0);
    };
    return TextData;
})();
var ArrayBufferData = (function () {
    function ArrayBufferData(value0) {
        this.value0 = value0;
    };
    ArrayBufferData.create = function (value0) {
        return new ArrayBufferData(value0);
    };
    return ArrayBufferData;
})();
var ArrayBufferViewData = (function () {
    function ArrayBufferViewData(value0) {
        this.value0 = value0;
    };
    ArrayBufferViewData.create = function (value0) {
        return new ArrayBufferViewData(value0);
    };
    return ArrayBufferViewData;
})();
var BlobData = (function () {
    function BlobData(value0) {
        this.value0 = value0;
    };
    BlobData.create = function (value0) {
        return new BlobData(value0);
    };
    return BlobData;
})();
var FormData = (function () {
    function FormData(value0) {
        this.value0 = value0;
    };
    FormData.create = function (value0) {
        return new FormData(value0);
    };
    return FormData;
})();
var DocumentData = (function () {
    function DocumentData(value0) {
        this.value0 = value0;
    };
    DocumentData.create = function (value0) {
        return new DocumentData(value0);
    };
    return DocumentData;
})();
var JsonData = (function () {
    function JsonData(value0) {
        this.value0 = value0;
    };
    JsonData.create = function (value0) {
        return new JsonData(value0);
    };
    return JsonData;
})();
var showResponseType = new Prelude.Show(function (_10) {
    if (_10 instanceof Default) {
        return "";
    };
    if (_10 instanceof ArrayBuffer) {
        return "arraybuffer";
    };
    if (_10 instanceof Blob) {
        return "blob";
    };
    if (_10 instanceof Document) {
        return "document";
    };
    if (_10 instanceof Json) {
        return "json";
    };
    if (_10 instanceof Text) {
        return "text";
    };
    if (_10 instanceof MozBlob) {
        return "moz-blob";
    };
    if (_10 instanceof MozChunkedText) {
        return "moz-chunked-text";
    };
    if (_10 instanceof MozChunkedArrayBuffer) {
        return "moz-chunked-arraybuffer";
    };
    throw new Error("Failed pattern match");
});
var showHttpMethod = new Prelude.Show(function (_9) {
    if (_9 instanceof GET) {
        return "GET";
    };
    if (_9 instanceof POST) {
        return "POST";
    };
    if (_9 instanceof PUT) {
        return "PUT";
    };
    if (_9 instanceof DELETE) {
        return "DELETE";
    };
    if (_9 instanceof PATCH) {
        return "PATCH";
    };
    if (_9 instanceof HEAD) {
        return "HEAD";
    };
    if (_9 instanceof OPTIONS) {
        return "OPTIONS";
    };
    if (_9 instanceof JSONP) {
        return "JSONP";
    };
    if (_9 instanceof HttpMethod) {
        return _9.value0;
    };
    throw new Error("Failed pattern match");
});

/**
 *  XXX should be encoded
 */
var setResponseType = function (rt) {
    return function (x) {
        return Data_DOM_Simple_Unsafe_Ajax.unsafeSetResponseType(x, Prelude.show(showResponseType)(rt));
    };
};
var send = function (_7) {
    return function (_8) {
        if (_7 instanceof NoData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSend(_8);
        };
        if (_7 instanceof TextData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSendWithPayload(_8, _7.value0);
        };
        if (_7 instanceof ArrayBufferData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSendWithPayload(_8, _7.value0);
        };
        if (_7 instanceof ArrayBufferViewData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSendWithPayload(_8, _7.value0);
        };
        if (_7 instanceof BlobData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSendWithPayload(_8, _7.value0);
        };
        if (_7 instanceof DocumentData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSendWithPayload(_8, _7.value0);
        };
        if (_7 instanceof FormData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSendWithPayload(_8, _7.value0);
        };
        if (_7 instanceof JsonData) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeSendWithPayload(_8, _7.value0);
        };
        throw new Error("Failed pattern match");
    };
};
var responseType = function (obj) {
    return function __do() {
        var _1 = Data_DOM_Simple_Unsafe_Ajax.unsafeResponseType(obj)();
        if (_1 === "") {
            return Default.value;
        };
        if (_1 === "arraybuffer") {
            return ArrayBuffer.value;
        };
        if (_1 === "blob") {
            return Blob.value;
        };
        if (_1 === "document") {
            return Document.value;
        };
        if (_1 === "json") {
            return Json.value;
        };
        if (_1 === "text") {
            return Text.value;
        };
        if (_1 === "moz-blob") {
            return MozBlob.value;
        };
        if (_1 === "moz-chunked-test") {
            return MozChunkedText.value;
        };
        if (_1 === "moz-chunked-arraybuffer") {
            return MozChunkedArrayBuffer.value;
        };
        throw new Error("Failed pattern match");
    };
};
var response = function (x) {
    var get = function (t) {
        return Prelude["<$>"](Control_Monad_Eff.functorEff)(Data_Function.runFn3(maybeFn)(NoData.value)(t))(Data_DOM_Simple_Unsafe_Ajax.unsafeResponse(x));
    };
    return function __do() {
        var _2 = responseType(x)();
        return (function () {
            if (_2 instanceof Default) {
                return get(TextData.create);
            };
            if (_2 instanceof ArrayBuffer) {
                return get(ArrayBufferData.create);
            };
            if (_2 instanceof Blob) {
                return get(BlobData.create);
            };
            if (_2 instanceof Document) {
                return get(DocumentData.create);
            };
            if (_2 instanceof Json) {
                return get(JsonData.create);
            };
            if (_2 instanceof Text) {
                return get(TextData.create);
            };
            if (_2 instanceof MozBlob) {
                return get(BlobData.create);
            };
            if (_2 instanceof MozChunkedText) {
                return get(TextData.create);
            };
            if (_2 instanceof MozChunkedArrayBuffer) {
                return get(ArrayBufferData.create);
            };
            throw new Error("Failed pattern match");
        })()();
    };
};
var readyState = function (x) {
    return function __do() {
        var _0 = Data_DOM_Simple_Unsafe_Ajax.unsafeReadyState(x)();
        if (_0 === 0) {
            return Unsent.value;
        };
        if (_0 === 1) {
            return Opened.value;
        };
        if (_0 === 2) {
            return HeadersReceived.value;
        };
        if (_0 === 3) {
            return Loading.value;
        };
        if (_0 === 4) {
            return Done.value;
        };
        throw new Error("Failed pattern match");
    };
};
var open = function (m) {
    return function (u) {
        return function (x) {
            return Data_DOM_Simple_Unsafe_Ajax.unsafeOpen(x, Prelude.show(showHttpMethod)(m), u);
        };
    };
};
var onReadyStateChange = function (f) {
    return function (x) {
        return Data_DOM_Simple_Unsafe_Ajax.unsafeOnReadyStateChange(x, f);
    };
};
var maybe = Data_Function.runFn3(maybeFn)(Data_Maybe.Nothing.value)(Data_Maybe.Just.create);
var getResponseHeader = function (k) {
    return function (x) {
        return Prelude["<$>"](Control_Monad_Eff.functorEff)(maybe)(Data_DOM_Simple_Unsafe_Ajax.unsafeGetResponseHeader(x, k));
    };
};
module.exports = {
    NoData: NoData, 
    TextData: TextData, 
    ArrayBufferData: ArrayBufferData, 
    ArrayBufferViewData: ArrayBufferViewData, 
    BlobData: BlobData, 
    FormData: FormData, 
    DocumentData: DocumentData, 
    JsonData: JsonData, 
    Default: Default, 
    ArrayBuffer: ArrayBuffer, 
    Blob: Blob, 
    Document: Document, 
    Json: Json, 
    Text: Text, 
    MozBlob: MozBlob, 
    MozChunkedText: MozChunkedText, 
    MozChunkedArrayBuffer: MozChunkedArrayBuffer, 
    GET: GET, 
    POST: POST, 
    PUT: PUT, 
    DELETE: DELETE, 
    PATCH: PATCH, 
    HEAD: HEAD, 
    OPTIONS: OPTIONS, 
    JSONP: JSONP, 
    HttpMethod: HttpMethod, 
    Unsent: Unsent, 
    Opened: Opened, 
    HeadersReceived: HeadersReceived, 
    Loading: Loading, 
    Done: Done, 
    overrideMimeType: overrideMimeType, 
    getResponseHeader: getResponseHeader, 
    getAllResponseHeaders: getAllResponseHeaders, 
    setRequestHeader: setRequestHeader, 
    statusText: statusText, 
    status: status, 
    responseText: responseText, 
    response: response, 
    responseType: responseType, 
    setResponseType: setResponseType, 
    send: send, 
    open: open, 
    onReadyStateChange: onReadyStateChange, 
    readyState: readyState, 
    makeXMLHttpRequest: makeXMLHttpRequest, 
    showHttpMethod: showHttpMethod, 
    showResponseType: showResponseType
};
